# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Abdullah-9862873/FlyRank-Week-1/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

> I pick the **Refresh / Content Opportunity Scoring** lane. The idea: recommend which content pages to refresh (rework) first. Over half of the pages in the starter data are already measured as declining in traffic, so ranking which one to refresh first adds real value. It also matches the trend and impressions questions I already explored, so I can reuse what I have learned.

In [1]:
import os
import pandas as pd

def locate_csv():
    roots = ['/content', './', '../', '../../data/raw', '/']
    for r in roots:
        for dirpath, _, files in os.walk(r):
            if 'content_refresh_anonymized.csv' in files:
                return os.path.join(dirpath, 'content_refresh_anonymized.csv')
    return None

path = locate_csv()
if path is None:
    raise FileNotFoundError('CSV not found anywhere. Upload content_refresh_anonymized.csv to Colab.')
print('Using:', path)

df = pd.read_csv(path)

declining_share = (df['trend_direction'] == 'down').mean()
declining_pages = (df['trend_direction'] == 'down').sum()
total_imp = df['impressions_90d'].sum()

print('Rows:', len(df))
print('Declining share:', round(declining_share * 100, 1), '%')
print('Declining pages:', declining_pages)
print('Total 90-day impressions:', total_imp)

Using: ../content_refresh_anonymized.csv
Rows: 30000
Declining share: 54.2 %
Declining pages: 16262
Total 90-day impressions: 156010989


## 2. The question: decision, action, cost of a wrong call

**Question:** Which content pages should a publisher refresh (rework) first?

**Decision it improves:** Which page to refresh next.

**Who acts on it:** An editor who picks the highest-ranked page from the list and reworks it.

**Cost of a wrong recommendation:** Two real costs. First, effort spent reworking a page that was already healthy. Second, missing a page that keeps losing traffic, so it falls for another cycle. Both waste time and lost impressions.

In [2]:
# This cell backs the decision framing with a base-rate check.
import os
import pandas as pd

def locate_csv():
    roots = ['/content', './', '../', '../../data/raw', '/']
    for r in roots:
        for dirpath, _, files in os.walk(r):
            if 'content_refresh_anonymized.csv' in files:
                return os.path.join(dirpath, 'content_refresh_anonymized.csv')
    return None

path = locate_csv()
if path is None:
    raise FileNotFoundError('CSV not found anywhere. Upload content_refresh_anonymized.csv to Colab.')
print('Using:', path)

df = pd.read_csv(path)

base_rate = (df['trend_direction'] == 'down').mean()
print('Base rate of declining pages:', round(base_rate * 100, 1), '%')
print('A random pick is close to a coin flip, so ranking has room to help.')

Using: ../content_refresh_anonymized.csv
Base rate of declining pages: 54.2 %
A random pick is close to a coin flip, so ranking has room to help.


## 3. Quick look at the data (2-3 real numbers)

> These numbers show why the refresh lane is worth the next weeks of work.
>
> - **54.2%** of pages are measured as declining (`trend_direction == 'down'`, 16,262 of 30,000).
> - **156,010,989** total 90-day impressions across all rows, so the decision matters at scale.
> - The base rate makes ranking worthwhile: a random pick on trend is close to a coin flip.

In [3]:
import os
import pandas as pd

def locate_csv():
    roots = ['/content', './', '../', '../../data/raw', '/']
    for r in roots:
        for dirpath, _, files in os.walk(r):
            if 'content_refresh_anonymized.csv' in files:
                return os.path.join(dirpath, 'content_refresh_anonymized.csv')
    return None

path = locate_csv()
if path is None:
    raise FileNotFoundError('CSV not found anywhere. Upload content_refresh_anonymized.csv to Colab.')
print('Using:', path)

df = pd.read_csv(path)

declining_share = (df['trend_direction'] == 'down').mean()
declining_pages = (df['trend_direction'] == 'down').sum()
total_imp = df['impressions_90d'].sum()

print('Rows:', len(df))
print('Declining share:', round(declining_share * 100, 1), '%')
print('Declining pages:', declining_pages)
print('Total 90-day impressions:', total_imp)

Using: ../content_refresh_anonymized.csv
Rows: 30000
Declining share: 54.2 %
Declining pages: 16262
Total 90-day impressions: 156010989


## 4. Careful words: what I can and can't claim

**What I can say:** The trend direction, the number of declining pages, and the total impressions are all measured from the data. The result is a ranking (decision support) for prioritising which page to refresh, not a certainty.

**What I can never claim:** That refreshing a page causes traffic to recover (that would need causal checks), a specific traffic number, or that I can predict Google's changes. I only rank which pages look most likely to be worth a refresh look first.

In [4]:
# Honest scope note for the write-up.
scope = (
    'Can claim: measured trend direction, measured impressions, a decision-support ranking. '
    'Cannot claim: causal proof that refresh improves traffic, or a prediction of Google.'
)
print(scope)

Can claim: measured trend direction, measured impressions, a decision-support ranking. Cannot claim: causal proof that refresh improves traffic, or a prediction of Google.


## Self-check

Before you submit, confirm each line honestly:

- [✓] Every section above is filled — markdown thinking AND the code that backs it
- [✓] The notebook runs top to bottom with no errors (Runtime → Run all)
- [✓] No client names, URLs, or private queries anywhere
- [✓] My claims use careful words: observed, measured, directional, decision-support
- [✓] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.